### PR TITLE
[tuner]: two new utility functions which are more friendly for c binding

### PR DESCRIPTION
**Pull request by bangtianliu**
Follow-up PR after https://github.com/iree-org/iree/pull/19124. 

TODO: add binding in this PR.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp [hunk 1]
```diff
@@ -1028,22 +1028,25 @@ std::optional<int> getGPUSubgroupSize(mlir::FunctionOpInterface func) {
...
-            return attr.getIntrinsic().getValue();
-          });
-      mmaAttributesMap[executableOp] = std::move(mmaIntrinsics);
-    }
-  });
-  return mmaAttributesMap;
+llvm::SmallVector<IREE::HAL::ExecutableVariantOp>
+getExecutableVariantOps(mlir::ModuleOp moduleOp) {
+  llvm::SmallVector<IREE::HAL::ExecutableVariantOp> executableVariantOps;
+  moduleOp.walk<WalkOrder::PreOrder>(
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]: Why did you change it to pre-order?

File: compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp [hunk 2]
```diff
@@ -1028,22 +1028,25 @@ std::optional<int> getGPUSubgroupSize(mlir::FunctionOpInterface func) {
...
-          target.getWgp().getMma(), [](IREE::GPU::MMAAttr attr) {
-            return attr.getIntrinsic().getValue();
-          });
-      mmaAttributesMap[executableOp] = std::move(mmaIntrinsics);
-    }
-  });
-  return mmaAttributesMap;
+llvm::SmallVector<IREE::HAL::ExecutableVariantOp>
+getExecutableVariantOps(mlir::ModuleOp moduleOp) {
+  llvm::SmallVector<IREE::HAL::ExecutableVariantOp> executableVariantOps;
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]:
```suggestion
SmallVector<IREE::HAL::ExecutableVariantOp>
getExecutableVariantOps(mlir::ModuleOp moduleOp) {
  SmallVector<IREE::HAL::ExecutableVariantOp> executableVariantOps;
```

File: compiler/src/iree/compiler/Codegen/Utils/GPUUtils.h [hunk 3]
```diff
@@ -207,13 +207,16 @@ IREE::GPU::TargetAttr getGPUTargetAttr(Operation *op);
...
 
-/// Returns a map of supported MMA intrinsic instructions based on the
-/// GPU target descriptions in `moduleOp`. Each entry in the map associates
-/// an `IREE::HAL::ExecutableVariantOp` with a vector of
-/// `IREE::GPU::MMAIntrinsic` attributes.
-llvm::SmallDenseMap<IREE::HAL::ExecutableVariantOp,
-                    SmallVector<IREE::GPU::MMAIntrinsic>>
-queryMMAIntrinsics(mlir::ModuleOp moduleOp);
+/// Returns all `IREE::HAL::ExecutableVariantOp` operations from the
+/// given `mlir::ModuleOp`,  ensuring they are returned in their original IR
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Utils/GPUUtils.h]:
```suggestion
/// given `mlir::ModuleOp`, ensuring they are returned in their original IR
```

File: compiler/src/iree/compiler/Codegen/Utils/GPUUtils.h [hunk 4]
```diff
@@ -207,13 +207,16 @@ IREE::GPU::TargetAttr getGPUTargetAttr(Operation *op);
...
+/// order.
+llvm::SmallVector<IREE::HAL::ExecutableVariantOp>
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Utils/GPUUtils.h]:
```suggestion
SmallVector<IREE::HAL::ExecutableVariantOp>
```

File: compiler/src/iree/compiler/Codegen/Utils/GPUUtils.h [hunk 5]
```diff
@@ -207,13 +207,16 @@ IREE::GPU::TargetAttr getGPUTargetAttr(Operation *op);
...
+getExecutableVariantOps(mlir::ModuleOp moduleOp);
+
+// Returns the MMA intrinsics associated with the given
+// `IREE::HAL::ExecutableVariantOp`.
+llvm::SmallVector<IREE::GPU::MMAIntrinsic>
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Utils/GPUUtils.h]:
```suggestion
SmallVector<IREE::GPU::MMAIntrinsic>
```

File: compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp (hunk 1 above)
Comment (bangtianliu) [reply compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]: Just to ensure it is in correct order.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TestLLVMGPUQueryMMAPass.cpp [hunk 6]
```diff
@@ -23,15 +23,16 @@ struct TestLLVMGPUQueryMMAPass final
     : impl::TestLLVMGPUQueryMMAPassBase<TestLLVMGPUQueryMMAPass> {
   void runOnOperation() override {
     ModuleOp moduleOp = getOperation();
-    llvm::SmallDenseMap<IREE::HAL::ExecutableVariantOp,
-                        SmallVector<IREE::GPU::MMAIntrinsic>>
-        mmaMap = queryMMAIntrinsics(moduleOp);
-    for (const auto &[op, mmaAttrs] : mmaMap) {
+    SmallVector<IREE::HAL::ExecutableVariantOp> executableVariantOps =
+        getExecutableVariantOps(moduleOp);
+    for (const auto &op : executableVariantOps) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TestLLVMGPUQueryMMAPass.cpp]: You can use values for IR constructs -- they are designed to be cheap to pass by value. Also use the actual type, since the type is not obvious based on the RHS only.

**Review by kuhar — APPROVED** [review]
LGTM % one nit
